File: backend/endpoints/library.py
import flask
from pydantic import BaseModel
from google.cloud import firestore

from backend.common import connect
from backend.common.app_access import require_access_level
from backend.common.database import DocumentsRef, LibraryRef
from backend.common.models import Document, Favorite
from backend.common.models import Document, Favorite, Vendor
from backend.endpoints.cache import cacheable_route
from onshape_api.endpoints.documents import ElementType
from onshape_api.endpoints.thumbnails import ThumbnailSize
from onshape_api.paths.instance_type import InstanceType


router = flask.Blueprint("library", __name__)


class InstancePathOut(BaseModel):
    documentId: str
    instanceId: str
    instanceType: InstanceType


class ElementPathOut(BaseModel):
    documentId: str
    instanceId: str
    instanceType: InstanceType
    elementId: str


class ElementOut(BaseModel):
    id: str
    documentId: str

    path: ElementPathOut
    name: str
    microversionId: str
    isVisible: bool
    isOpenComposite: bool
    supportsFasten: bool
    elementType: ElementType
    thumbnailUrls: dict[ThumbnailSize, str]
    configurationId: str | None
    vendors: list[Vendor]


class DocumentOut(BaseModel):
    id: str

    path: InstancePathOut
    name: str
    sortAlphabetically: bool
    thumbnailUrls: dict[ThumbnailSize, str]
    elementOrder: list[str]


class LibraryOut(BaseModel):
    """A model representing a library that is sent to the frontend."""

    documentOrder: list[str]
    documents: dict[str, DocumentOut]
    elements: dict[str, ElementOut]


@cacheable_route(router, connect.library_route())
def get_library(**kwargs):
    library_ref = connect.get_library_ref()
    return build_library_out(library_ref).model_dump_json(exclude_none=True)


def build_library_out(library_ref: LibraryRef) -> LibraryOut:
    documents, elements = build_documents_out(library_ref.documents)
    return LibraryOut(
        documentOrder=library_ref.documents.keys(),
        documents=documents,
        elements=elements,
    )


def build_documents_out(
    documents_ref: DocumentsRef,
) -> tuple[dict[str, DocumentOut], dict[str, ElementOut]]:
    documents_out: dict[str, DocumentOut] = {}
    elements_out: dict[str, ElementOut] = {}

    for document_ref in documents_ref.list():
        document: Document = document_ref.get()
        document_id = document_ref.id
        documents_out[document_id] = DocumentOut(
            id=document_id,
            path=InstancePathOut(
                documentId=document_id,
                instanceId=document.instanceId,
                instanceType=InstanceType.VERSION,
            ),
            name=document.name,
            sortAlphabetically=document.sortAlphabetically,
            thumbnailUrls=document.thumbnailUrls,
            elementOrder=document_ref.elements.keys(),
        )

        for element_ref in document_ref.elements.list():
            element = element_ref.get()
            elements_out[element_ref.id] = ElementOut(
                id=element_ref.id,
                documentId=document_id,
                path=ElementPathOut(
                    documentId=document_id,
                    instanceId=document.instanceId,
                    instanceType=InstanceType.VERSION,
                    elementId=element_ref.id,
                ),
                name=element.name,
                isVisible=element.isVisible,
                isOpenComposite=element.isOpenComposite,
                supportsFasten=element.fastenInfo != None,
                thumbnailUrls=element.thumbnailUrls,
                microversionId=element.microversionId,
                elementType=element.elementType,
                configurationId=element.configurationId,
                vendors=element.vendors,
            )

    return (documents_out, elements_out)


@cacheable_route(router, "/search-db" + connect.library_route())
def get_search_db(**kwargs):
    library_ref = connect.get_library_ref()
    library = library_ref.get()
    return {"searchDb": library.searchDb}


class FavoriteOut(BaseModel):
    id: str
    defaultConfiguration: dict[str, str] | None


class LibraryUserDataOut(BaseModel):
    favorites: dict[str, FavoriteOut]
    favoriteOrder: list[str]


@router.get("/library-user-data" + connect.library_route() + connect.user_path_route())
def get_library_user_data(**kwargs):
    user_id = connect.get_route_user_path().user_id
    library_ref = connect.get_library_ref()

    favorites_ref = library_ref.user_data.user_data(user_id).favorites
    favorites: dict[str, FavoriteOut] = {}
    for favorite_ref in favorites_ref.list():
        favorite = favorite_ref.get()
        favorites[favorite_ref.id] = FavoriteOut(
            id=favorite_ref.id, defaultConfiguration=favorite.defaultConfiguration
        )

    return LibraryUserDataOut(
        favorites=favorites, favoriteOrder=favorites_ref.keys()
    ).model_dump_json(exclude_none=True)


@router.post("/favorites" + connect.library_route() + connect.user_path_route())
def add_favorite(**kwargs):
    library_ref = connect.get_library_ref()
    user_path = connect.get_route_user_path()
    element_id = connect.get_query_param("elementId")
    default_configuration = connect.get_optional_body_arg("defaultConfiguration")

    # Add it to favorite-order
    user_ref = library_ref.user_data.user_data(user_path.user_id)
    user_ref.favorites.add(
        element_id, Favorite(defaultConfiguration=default_configuration)
    )

    return {"success": True}


@router.delete("/favorites" + connect.library_route() + connect.user_path_route())
def remove_favorite(**kwargs):
    library_ref = connect.get_library_ref()
    user_path = connect.get_route_user_path()
    element_id = connect.get_query_param("elementId")

    user_ref = library_ref.user_data.user_data(user_path.user_id)
    user_ref.favorites.remove(element_id)

    return {"success": True}


@router.post("/favorite-order" + connect.library_route() + connect.user_path_route())
def set_favorite_order(**kwargs):
    """Sets the order of the current user's favorites."""
    library_ref = connect.get_library_ref()
    user_path = connect.get_route_user_path()
    favorite_order = connect.get_body_arg("favoriteOrder")

    user_ref = library_ref.user_data.user_data(user_path.user_id)
    user_ref.favorites.set_order(favorite_order)

    return {"success": True}


@router.post(
    "/default-configuration" + connect.library_route() + connect.user_path_route()
)
def update_default_configuration(**kwargs):
    library_ref = connect.get_library_ref()
    user_path = connect.get_route_user_path()
    favorite_id = connect.get_body_arg("favoriteId")
    default_configuration = connect.get_body_arg("defaultConfiguration")

    favorite_ref = library_ref.user_data.user_data(
        user_path.user_id
    ).favorites.favorite(favorite_id)

    favorite_ref.update({"defaultConfiguration": default_configuration})

    return {"success": True}


@router.post("/library-version" + connect.library_route())
@require_access_level()
def push_library_version(**kwargs):
    """
    Invalidates all CDN caching by pushing a new version of a library.
    """
    library_ref = connect.get_library_ref()
    new_search_db = connect.get_body_arg("searchDb")

    library_ref.update(
        {"searchDb": new_search_db, "cacheVersion": firestore.Increment(1)}
    )
    return {"success": True}

File: old/src/api/library.ts
import { createServerFn } from "@tanstack/react-start";
import { FieldValue } from "firebase-admin/firestore";
import {
    Library,
    ElementType,
    ThumbnailUrls,
    Vendor
} from "../api-utils/client-models";
import { OAuthApi } from "../onshape-api/client/oauth-api";
import {
    AccessLevel,
    getAccessLevel,
    getUserId
} from "../onshape-api/endpoints/users";
import { getLibrary, LibraryRef } from "../connect/library";
import { getOnshapeApi } from "../routes/auth/-auth.server";

export interface DocumentPathOut {
    documentId: string;
    instanceId: string;
    instanceType: "v";
}

export interface ElementPathOut extends DocumentPathOut {
    elementId: string;
}

export interface ElementOut {
    id: string;
    documentId: string;
    path: ElementPathOut;
    name: string;
    microversionId: string;
    isVisible: boolean;
    isOpenComposite: boolean;
    supportsFasten: boolean;
    elementType: ElementType;
    thumbnailUrls: ThumbnailUrls;
    configurationId?: string;
    vendors: Vendor[];
}

export interface DocumentOut {
    id: string;
    path: DocumentPathOut;
    name: string;
    sortAlphabetically: boolean;
    thumbnailUrls: ThumbnailUrls;
    elementOrder: string[];
}

export interface LibraryOut {
    documentOrder: string[];
    documents: Record<string, DocumentOut>;
    elements: Record<string, ElementOut>;
}

export interface FavoriteOut {
    id: string;
    defaultConfiguration: Record<string, string> | undefined;
}

export interface LibraryUserDataOut {
    favorites: Record<string, FavoriteOut>;
    favoriteOrder: string[];
}

async function getAppAccessLevel(onshapeApi: OAuthApi): Promise<AccessLevel> {
    const override = process.env.ACCESS_LEVEL_OVERRIDE;
    if (override) return override as AccessLevel;

    const adminTeam = process.env.ADMIN_TEAM;
    if (!adminTeam)
        throw new Error(
            "ADMIN_TEAM or ACCESS_LEVEL_OVERRIDE must be configured"
        );

    return getAccessLevel(onshapeApi, adminTeam);
}

async function requireAccess(
    onshapeApi: OAuthApi,
    required: AccessLevel = AccessLevel.EDITOR
): Promise<void> {
    const level = await getAppAccessLevel(onshapeApi);
    if (
        required === AccessLevel.EDITOR &&
        (level === AccessLevel.EDITOR || level === AccessLevel.ADMIN)
    )
        return;
    if (required === AccessLevel.ADMIN && level === AccessLevel.ADMIN) return;
    throw new Error("Insufficient permissions");
}

export async function buildLibraryOut(
    libraryRef: LibraryRef
): Promise<LibraryOut> {
    const library = await libraryRef.get();
    if (!library) throw new Error("Library not found");

    const documents: Record<string, DocumentOut> = {};
    const elements: Record<string, ElementOut> = {};

    await Promise.all(
        library.documentOrder.map(async (documentId) => {
            const docRef = libraryRef.documents.document(documentId);
            const document = await docRef.get();
            if (!document) return;

            documents[documentId] = {
                id: documentId,
                path: {
                    documentId,
                    instanceId: document.instanceId,
                    instanceType: "v"
                },
                name: document.name,
                sortAlphabetically: document.sortAlphabetically,
                thumbnailUrls: document.thumbnailUrls as ThumbnailUrls,
                elementOrder: document.elementOrder
            };

            await Promise.all(
                document.elementOrder.map(async (elementId) => {
                    const element = await docRef.elements
                        .element(elementId)
                        .get();
                    if (!element) return;

                    elements[elementId] = {
                        id: elementId,
                        documentId,
                        path: {
                            documentId,
                            instanceId: document.instanceId,
                            instanceType: "v",
                            elementId
                        },
                        name: element.name,
                        microversionId: element.microversionId,
                        isVisible: element.isVisible,
                        isOpenComposite: element.isOpenComposite,
                        supportsFasten: element.fastenInfo !== undefined,
                        elementType: element.elementType,
                        thumbnailUrls: element.thumbnailUrls as ThumbnailUrls,
                        configurationId: element.configurationId,
                        vendors: element.vendors
                    };
                })
            );
        })
    );

    return { documentOrder: library.documentOrder, documents, elements };
}

// --- Server functions ---

export const fetchLibraryData = createServerFn()
    .inputValidator((data: { library: Library }) => data)
    .handler(async ({ data: { library } }) => {
        return buildLibraryOut(getLibrary(library));
    });

export const fetchSearchDb = createServerFn()
    .inputValidator((data: { library: Library }) => data)
    .handler(async ({ data: { library } }) => {
        const lib = await getLibrary(library).get();
        return { searchDb: lib?.searchDb };
    });

export const fetchLibraryUserData = createServerFn()
    .inputValidator((data: { library: Library }) => data)
    .handler(async ({ data: { library } }): Promise<LibraryUserDataOut> => {
        const onshapeApi = await getOnshapeApi();
        const userId = await getUserId(onshapeApi);

        const favoritesRef =
            getLibrary(library).userData.userDataFor(userId).favorites;
        const favoriteEntries = await favoritesRef.list();

        const favorites: Record<string, FavoriteOut> = {};
        for (const { id, data } of favoriteEntries) {
            favorites[id] = {
                id,
                defaultConfiguration: data.defaultConfiguration
            };
        }

        return {
            favorites,
            favoriteOrder: favoriteEntries.map(({ id }) => id)
        };
    });

export const addFavorite = createServerFn()
    .inputValidator(
        (data: {
            library: Library;
            elementId: string;
            defaultConfiguration?: Record<string, string>;
        }) => data
    )
    .handler(async ({ data: { library, elementId, defaultConfiguration } }) => {
        const onshapeApi = await getOnshapeApi();
        const userId = await getUserId(onshapeApi);

        await getLibrary(library)
            .userData.userDataFor(userId)
            .favorites.add(elementId, {
                defaultConfiguration
            });

        return { success: true };
    });

export const removeFavorite = createServerFn()
    .inputValidator((data: { library: Library; elementId: string }) => data)
    .handler(async ({ data: { library, elementId } }) => {
        const onshapeApi = await getOnshapeApi();
        const userId = await getUserId(onshapeApi);

        await getLibrary(library)
            .userData.userDataFor(userId)
            .favorites.remove(elementId);

        return { success: true };
    });

/** Sets the order of the current user's favorites. */
export const setFavoriteOrder = createServerFn()
    .inputValidator(
        (data: { library: Library; favoriteOrder: string[] }) => data
    )
    .handler(async ({ data: { library, favoriteOrder } }) => {
        const onshapeApi = await getOnshapeApi();
        const userId = await getUserId(onshapeApi);

        await getLibrary(library)
            .userData.userDataFor(userId)
            .update({ favoriteOrder });

        return { success: true };
    });

export const updateDefaultConfiguration = createServerFn()
    .inputValidator(
        (data: {
            library: Library;
            favoriteId: string;
            defaultConfiguration: Record<string, string>;
        }) => data
    )
    .handler(
        async ({ data: { library, favoriteId, defaultConfiguration } }) => {
            const onshapeApi = await getOnshapeApi();
            const userId = await getUserId(onshapeApi);

            await getLibrary(library)
                .userData.userDataFor(userId)
                .favorites.favorite(favoriteId)
                .update({ defaultConfiguration });

            return { success: true };
        }
    );

/**
 * Invalidates all CDN caching by incrementing the library's cache version.
 */
export const pushLibraryVersion = createServerFn({ method: "POST" })
    .inputValidator((data: { library: Library; searchDb: string }) => data)
    .handler(async ({ data: { library, searchDb } }) => {
        const onshapeApi = await getOnshapeApi();
        await requireAccess(onshapeApi);

        const libraryRef = getLibrary(library);
        await libraryRef.ref.update({
            searchDb,
            cacheVersion: FieldValue.increment(1)
        });

        return { success: true };
    });
